pkg/apis/batch/validation/validation.go

```diff
@@ -99,12 +105,18 @@ func ValidateGeneratedSelector(obj *batch.Job) field.ErrorList {
 	// have placed certain labels on the pod, but this could have failed if
 	// the user added coflicting labels.  Validate that the expected
 	// generated ones are there.
+	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), controllerUid, string(obj.UID))...)
+	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), jobName, string(obj.Name))...)
+	// Labels created by the Kubernetes project should have a Kubernetes prefix.
+	// These labels are set due to legacy reasons.
+	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), jobNameLegacy, string(obj.Name))...)
+	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), controllerUidLegacy, string(obj.UID))...)
 
-	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), "controller-uid", string(obj.UID))...)
-	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), "job-name", string(obj.Name))...)
 	expectedLabels := make(map[string]string)
-	expectedLabels["controller-uid"] = string(obj.UID)
-	expectedLabels["job-name"] = string(obj.Name)
+	expectedLabels[controllerUidLegacy] = string(obj.UID)
+	expectedLabels[jobNameLegacy] = string(obj.Name)
+	expectedLabels[controllerUid] = string(obj.UID)
+	expectedLabels[jobName] = string(obj.Name)
```

Comment:
For that change to be backwards compatible, we need to only validate the Job name with the label value rules only when using a generated selector.

Reply:
If you have a Job with a long name (and not manual selector) you will try to stick that long name into the label value.  Given that this is happening in validation, AFTER label validation, you will end up inserting corrupted (too long) labels and (I think) storing them!

The next update should fail validation in a very confusing way.

If you properly set these labels in PrepareForCreate() or even in pkg/apis/batch/v1/defaults.go, then validation would fail - still with a bad error, but at least it would fail.  You could do better by checking the `manualSelector` flag and then failing if the name is too long, with a good error.

Please also add a testcase of creating a Job with a too-long name.

Reply:
Do you think this can also be done as a followup? 

I don't think this is related to adding these labels as we aren't checking the max length of label plus name.

Reply:
sure, this is not a new failure mode - it exists today.

Can you please file another issue to capture? 

Thanks

Reply:
https://github.com/kubernetes/kubernetes/pull/116146